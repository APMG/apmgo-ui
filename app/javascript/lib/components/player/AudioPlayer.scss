@import '../../styles/helpers';

.player {
  height: $player-height-mobile;
  padding: 20px;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: $z-player;
  text-align: center;
  color: $color-white;
  background-color: $color-grey-dark;

  @include media($media-small-tablet) {
    height: $player-height-small-tablet;
  }

  @include media($media-tablet) {
    height: $player-height-tablet;
  }
}

.title {
  margin-bottom: 15px;
  font-size: 1.8rem;
  font-weight: 700;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  @include media($media-small-tablet) {
    font-size: 2.2rem;
    margin-bottom: 20px;
  }

  @include media($media-tablet) {
    font-size: 2.6rem;
    margin-bottom: 30px;
  }
}

.controls {
  margin-bottom: 10px;
}

.playPause {
  margin-bottom: 10px;
}

.playPauseBtn {
  height: 48px;
  width: 48px;
  margin: 0 60px;
  padding: 0 4px;
  border: 0;
  fill: $color-white;
  vertical-align: middle;
  background: transparent;
  transition: background-color $transition;

  &:focus {
    outline: 0;
    background-color: rgba($color-white, .1);
  }
}

.skipBtn {
  height: 35px;
  width: 35px;
  position: relative;
  background: transparent;
  border: 0;
  padding: 12px 10px 8px;
  font-size: 1.3rem;
  vertical-align: middle;
  fill: $color-white;
  transition: background-color $transition;

  &:focus {
    outline: 0;
    background-color: rgba($color-white, .1);
  }

  svg {
    pointer-events: none;
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
  }
}

.skipBack {
  composes: skipBtn;
}

.skipFwd {
  composes: skipBtn;
}

.progress {
  margin-bottom: 15px;

  @include media($media-small-tablet) {
    margin-bottom: 20px;
  }
}

.timestamps {
  @include clearfix();
  margin-bottom: 10px;
}

.currentTime {
  float: left;
}

.duration {
  float: right;
}

.scrubber {
  width: 100%;
}

.volume {
  display: none;
}
