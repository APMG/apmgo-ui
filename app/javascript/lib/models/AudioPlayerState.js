// @flow
type AudioPlayerStateParams = {
  currentTrackId?: number,
  paused?: boolean,
  muted?: boolean,
  currentTime?: number,
  updateAudioElementTime?: boolean, // only present when explicitly set, signals need to update the time on the audio element
  canPlay?: boolean,
  volume?: number,
  duration?: number,
  isDragging?: boolean,
  playWhenCan?: boolean
}

export default class AudioPlayerState {
  currentTrackId: number | typeof undefined
  paused: boolean
  muted: boolean
  currentTime: number
  updateAudioElementTime: boolean
  canPlay: boolean
  volume: number
  duration: number
  isDragging: boolean
  playWhenCan: boolean

  constructor (params: AudioPlayerStateParams) {
    this.currentTrackId = params.currentTrackId
    this.paused = !!params.paused
    this.muted = !!params.muted
    this.currentTime = params.currentTime || 0
    this.updateAudioElementTime = params.updateAudioElementTime || false
    this.canPlay = params.canPlay || false
    this.duration = params.duration || 0
    this.isDragging = params.isDragging || false
    this.playWhenCan = params.playWhenCan || false

    if (params.volume === 0) {
      this.volume = 0
    } else {
      this.volume = params.volume || 1
    }
  }

  setCurrentTrackId (track: number) {
    return this._make({
      currentTrackId: track
    })
  }

  setTime (time: number) {
    return this._make({currentTime: time})
  }

  // If chaining AudioPlayerState updates,
  // this method must be called last
  // or it will be overwritten
  // because its value is not preserved in _make({})
  instanceUpdatesAudioElementTime () {
    return this._make({updateAudioElementTime: true})
  }

  setCanPlay (canPlay: boolean) {
    return this._make({canPlay: canPlay})
  }

  play (currentTrackId?: number) {
    currentTrackId = currentTrackId || this.currentTrackId

    return this._make({
      paused: false,
      currentTrackId: currentTrackId
    })
  }

  pause () {
    return this._make({paused: true})
  }

  mute () {
    return this._make({muted: true})
  }

  unmute () {
    return this._make({muted: false})
  }

  setVolume (volume: number) {
    return this._make({volume: volume})
  }

  setDuration (duration: number) {
    return this._make({duration: duration})
  }

  setPlayWhenCan (playWhenCan: boolean) {
    return this._make({playWhenCan: playWhenCan})
  }

  _make (params: {}): AudioPlayerState {
    let newParams = {
      currentTrackId: this.currentTrackId,
      paused: this.paused,
      muted: this.muted,
      currentTime: this.currentTime,
      canPlay: this.canPlay,
      volume: this.volume,
      duration: this.duration,
      isDragging: this.isDragging,
      playWhenCan: this.playWhenCan,
      // note doUpdateTime is not here by default,
      // it must be set explicitly, meaning that if
      // methods are chained, it will be overwritten
      // if instanceUpdatesAudioElementTime() is not called last
      ...params
    }

    return new AudioPlayerState(newParams)
  }
}
