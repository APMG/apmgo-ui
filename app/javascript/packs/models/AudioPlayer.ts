import { AudioPlayerType } from '../redux/types'

export default class AudioPlayer implements AudioPlayerType {

  item_id: number
  paused: boolean

  constructor(item_id: number) {
    this.item_id = item_id
    this.paused = true;
  }
}