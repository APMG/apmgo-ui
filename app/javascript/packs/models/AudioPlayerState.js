// @flow

type AudioPlayerStateParams = {
  currentTrackId?: number,
  paused?: boolean,
  muted?: boolean,
  currentTime?: number
}

export default class AudioPlayerState {
  currentTrackId: number | typeof undefined
  paused: boolean
  muted: boolean
  currentTime: number

  constructor(params: AudioPlayerStateParams) {

    this.currentTrackId = params.currentTrackId
    this.paused = !!params.paused
    this.muted = !!params.muted
    this.currentTime = params.currentTime || 0
  }

  setCurrentTrackId(track: number) {
    return this._make({
      currentTrackId: track
     })
  }

  setTime(time: number) {
    return this._make({currentTime: time})
  }

  play(currentTrackId?: number) {
    let params : {paused: boolean, currentTrackId?: number} = { paused: false }

    if (currentTrackId) {
      params.currentTrackId = currentTrackId
    }

    return this._make(params)
  }

  pause() {
    return this._make({paused: true})
  }

  mute() {
    return this._make({muted: true})
  }

  unmute() {
    return this._make({muted: false})
  }

  _make(params: {}) : AudioPlayerState {
    let newParams = {
      currentTrackId: this.currentTrackId,
      paused: this.paused,
      muted: this.muted,
      currentTime: this.currentTime,
      ...params
    }
    return new AudioPlayerState(newParams)
  }
}
