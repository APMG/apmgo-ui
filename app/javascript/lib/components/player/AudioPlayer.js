// @flow
import React from 'react';
import { connect, dispatch } from 'react-redux'

import AudioPlayerState from '../../models/AudioPlayerState';
import type { PlaylistItemType } from '../../redux/types';
import PlayPauseButton from './PlayPauseButton'
import MuteButton from './MuteButton'
import TimeScrubber from './TimeScrubber'
import TimeKeeper from './TimeKeeper'
import VolumeSlider from './VolumeSlider'
import PlayTimeDisplay from './PlayTimeDisplay'

import {
  audioMetaDataLoaded,
  audioCanPlay,
  playClick,
  pauseClick,
  muteClick,
  unmuteClick,
  updatePlayTime,
  timeScrubberChange,
  volumeChange,
  changeTrack
} from '../../redux/audio-player'

import { archivePlaylistItem } from '../../redux/playlist'

type AudioPlayerProps = {
  item: PlaylistItemType,
  audioPlayer: AudioPlayerState,
  audioCanPlay: () => {},
  audioMetaDataLoaded: (duration: number) => {},
  pause: () => {},
  play: () => {},
  mute: () => {},
  unmute: () => {},
  updatePlayTimeTimeKeeper: () => {},
  updatePlayTimeTimeScrubber: () => {},
  timeScrubberChange: () => {},
  updateVolume: () => {},
  onEnded: () => {},
  changeTrack: (item: PlaylistItemType) => {}
}

export class AudioPlayerPresenter extends React.Component {

  audio: HTMLAudioElement
  props: AudioPlayerProps

  constructor(props: AudioPlayerProps) {
    super(props)
  }

  componentDidMount() {
    // by now, this.audio is an HTMLAudioElement ref
    // it gets set when the component renders...
    // see the audio element's `ref` attribute

    this.audio.src = this.props.item.attributes.audio_url
    this.audio.oncanplay = this.props.audioCanPlay,
    this.audio.onloadedmetadata = this.metaDataLoaded.bind(this)
    this.audio.onended = this.props.onEnded

    // So, these are being set here to make the component testable
    // the audio element does not get rendered in tests
    // so that logic needed to be moved out of the render function
    // to here, where it can be overridden by a test class
  }

  componentWillReceiveProps(newProps: AudioPlayerProps) {

    this._setPlayPaused(newProps.audioPlayer.paused)
    this.audio.muted = newProps.audioPlayer.muted
    this.audio.volume = newProps.audioPlayer.volume

    if(newProps.item.id !== this.props.item.id) {
      this.audio.src = newProps.item.attributes.audio_url
      if(this.audio.canPlay && !this.props.audioPlayer.paused) {
        this.audio.play()
      }
    }

    if(newProps.item.id !== newProps.audioPlayer.currentTrackId) {
      this.props.changeTrack(newProps.item)
    }

    if(newProps.audioPlayer.updateAudioElementTime) {
      this.audio.currentTime = newProps.audioPlayer.currentTime
    }
  }

  _setPlayPaused(paused: boolean) {
    if (paused !== this.audio.paused) {
      if (paused) {
        this.audio.pause()
      } else {
        this.audio.play()
      }
    }
  }

  metaDataLoaded(a:any,b:any,c:any) {
    this.audio.currentTime = this.props.item.attributes.playtime
    this.props.audioMetaDataLoaded(this.audio.duration)
  }

  audioRefCallback(ref:any) {
    if (!this.audio) {
      this.audio = ref
    }
  }

  render() {
    return (
      <div>
        <h3>
            { this.props.item
              ? 'Now playing: ' + this.props.item.attributes.audio_title
              : 'Loading ...' }
        </h3>

        <audio
          ref={ (ref) => this.audio=ref }>
        </audio>

        <PlayPauseButton
          paused={this.props.audioPlayer.paused}
          canPlay={this.props.audioPlayer.canPlay}
          pause={this.props.pause}
          play={this.props.play}
        />
        <MuteButton
          muted={this.props.audioPlayer.muted}
          mute={this.props.mute}
          unmute={this.props.unmute}
        />

        <h3>Time Control</h3>
        <TimeKeeper // this component is invisible
          audio={ this.audio }
          updatePlayTime={this.props.updatePlayTimeTimeKeeper}
        />
        <PlayTimeDisplay
          currentTime={this.props.audioPlayer.currentTime}
        />
        <TimeScrubber
          updatePlayTime={this.props.updatePlayTimeTimeScrubber}
          timeScrubberChange={this.props.timeScrubberChange}
          paused={this.props.audioPlayer.paused}
          duration={this.props.audioPlayer.duration}
          currentTime={this.props.audioPlayer.currentTime}
        />

        <h3>Volume Control</h3>
        <VolumeSlider
          volume={this.props.audioPlayer.volume}
          updateVolume={this.props.updateVolume}
        />
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch: dispatch, ownProps: AudioPlayerProps) => {
  return {
    audioCanPlay: () => {
      dispatch(audioCanPlay())
    },
    audioMetaDataLoaded: (duration: number) => {
      dispatch(audioMetaDataLoaded(duration))
    },
    play: () : void => {
      dispatch(playClick(ownProps.item.id))
    },
    pause: (): void => {
      dispatch(pauseClick(ownProps.item.id))
    },
    mute: () => {
      dispatch(muteClick())
    },
    unmute: () => {
      dispatch(unmuteClick())
    },
    updatePlayTimeTimeKeeper: (currentTime: number) => {
      dispatch(updatePlayTime(ownProps.item.id, Math.ceil(currentTime)))
    },
    updatePlayTimeTimeScrubber: (currentTime: number) => {
      dispatch(updatePlayTime(ownProps.item.id, Math.ceil(currentTime)))
    },
    timeScrubberChange: (currentTime: number) => {
      dispatch(timeScrubberChange(ownProps.item.id, Math.ceil(currentTime)))
    },
    updateVolume: (volume: number) => {
      dispatch(volumeChange(volume))
    },
    onEnded: () => {
      dispatch(archivePlaylistItem(ownProps.item))
    },
    changeTrack: ((item: PlaylistItemType) => {
      dispatch(changeTrack(item))
    })
  }
}

export const mapStateToProps = (newState: any) => {
  return {
    audioPlayer: newState.audioPlayer
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AudioPlayerPresenter)
