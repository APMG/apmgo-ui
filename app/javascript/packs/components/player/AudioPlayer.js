// @flow
import React from 'react';
import { connect, dispatch } from 'react-redux'

import AudioPlayerState from '../../models/AudioPlayerState';
import type { PlaylistItemType } from '../../redux/types';
import PlayPauseButton from './PlayPauseButton'
import MuteButton from './MuteButton'
import TimeScrubber from './TimeScrubber'
import TimeKeeper from './TimeKeeper'
import VolumeSlider from './VolumeSlider'
import PlayTimeDisplay from './PlayTimeDisplay'
import { audioMetaDataLoaded, audioCanPlay } from '../../redux/audio-player'

type AudioPlayerProps = {
  item: PlaylistItemType,
  playerState: AudioPlayerState,
  audioCanPlay: () => {},
  audioMetaDataLoaded: () => {}
}

export class AudioPlayerPresenter extends React.Component {

  audio: HTMLAudioElement
  props: AudioPlayerProps

  constructor(props: AudioPlayerProps) {
    super(props)
  }

  componentDidMount() {
    // by now, audio is an HTMLAudioElement ref
    this.audio.src = this.props.item.attributes.audio_url
  }

  componentWillReceiveProps(newProps: AudioPlayerProps) {
    this._setPlayPaused(newProps.playerState.paused)
    this.audio.muted = newProps.playerState.muted
    this.audio.volume = newProps.playerState.volume

    if(newProps.item.id !== this.props.item.id) {
      this.audio.src = newProps.item.attributes.audio_url
    }

    if(newProps.playerState.updateAudioElementTime) {
      this.audio.currentTime = newProps.playerState.currentTime
    }
  }

  _setPlayPaused(paused: boolean) {
    if (paused !== this.audio.paused) {
      if (paused) {
        this.audio.pause()
      } else {
        this.audio.play()
      }
    }
  }

  metaDataLoaded(a:any,b:any,c:any) {
    this.audio.currentTime = this.props.item.attributes.playtime
    this.props.audioMetaDataLoaded(this.audio.duration)
  }

  render() {
    return (
      <div>
        <h2>My Custom Playlist</h2>
        <h3>
            {this.props.item
              ? 'Now playing: ' + this.props.item.attributes.audio_title
              : 'Loading ...'}
        </h3>

        <audio
          ref={(ref) => this.audio = ref }
          onCanPlay={this.props.audioCanPlay}
          onLoadedMetadata={this.metaDataLoaded.bind(this)}>
        </audio>

        <PlayPauseButton />
        <MuteButton />

        <h3>Time Control</h3>
        <TimeKeeper // this component is invisible
          audio={ this.audio }
          item_id={this.props.item.id}
        />
        <PlayTimeDisplay currentTime={this.props.playerState.currentTime} />
        <TimeScrubber
          item_id={ this.props.item.id }
          currentTime={ this.props.playerState.currentTime }
          paused={ this.props.playerState.paused }
        />

        <h3>Volume Control</h3>
        <VolumeSlider />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    audioCanPlay: () => {
      dispatch(audioCanPlay())
    },
    audioMetaDataLoaded: (duration) => {
      dispatch(audioMetaDataLoaded(duration))
    }
  }
}

const mapStateToProps = (newState) => {
  return {
    playerState: Object.assign({}, newState.audioPlayer)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AudioPlayerPresenter)
