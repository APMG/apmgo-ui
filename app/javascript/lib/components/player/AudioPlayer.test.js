import React from 'react'
import connect from 'react-redux'
import { mount, shallow } from 'enzyme'
import AudioPlayer, { AudioPlayerPresenter, mapStateToProps, mapDispatchToProps } from './AudioPlayer'
import {
  audioCanPlay as audioCanPlayAction,
  audioMetaDataLoaded as audioMetaDataLoadedAction,
  playClick,
  pauseClick,
  muteClick,
  unmuteClick,
  updatePlayTime,
  timeScrubberChange as timeScrubberChangeAction,
  volumeChange
} from '../../redux/audio-player'
import { getSnapshotJson, itemFixtures, getMockStore, getWrappedComponent } from '../../redux/__tests__/mock-initial-state'
import MockableAudio from '../../redux/__tests__/mock-audio'
import AudioPlayerState from '../../models/AudioPlayerState'

let audioMockRef,
    getAudioMockRef = () => audioMockRef

class TestAudioPlayerPresenter extends AudioPlayerPresenter {

  componentDidMount() {
    this.audio = getAudioMockRef()
    super.componentDidMount()
  }
  componentWillReceiveProps(newProps) {
    this.audio = getAudioMockRef()
    super.componentWillReceiveProps(newProps)
  }
}

describe('AudioPlayer Component Test', () => {

  describe('Presenter', () => {

      let testProps

      beforeEach(() => {
        audioMockRef = {
          play: jest.fn(() => audioMockRef.paused = false),
          pause: jest.fn(() => audioMockRef.paused = true)
        },
        testProps = {
          item: itemFixtures[0],
          audioCanPlay: jest.fn(),
          audioMetaDataLoaded: jest.fn(),
          onEnded: jest.fn(),
          changeTrack: jest.fn(),
          audioRefSet: jest.fn((ref, context) => context.audio = ref ),
          audioPlayer: {volume: 1}
        }
      })
    it('Renders', () => {
      let tree = getSnapshotJson(<TestAudioPlayerPresenter {...testProps} />)
      expect(tree).toMatchSnapshot()
    })

    it('Passes Functions To Audio Property', () => {
      let wrapper = mount(getWrappedComponent(<TestAudioPlayerPresenter { ...testProps} />, getMockStore()))

      wrapper.render()

      expect(getAudioMockRef().src).toEqual(itemFixtures[0].attributes.audio_url)
      expect(audioMockRef.oncanplay).toEqual(testProps.audioCanPlay)

      audioMockRef.oncanplay()
      audioMockRef.onloadedmetadata()
      audioMockRef.onended()

      expect(testProps.audioMetaDataLoaded).toHaveBeenCalled()
      expect(testProps.audioCanPlay).toHaveBeenCalled()
      expect(testProps.onEnded).toHaveBeenCalled()
    })

    it('Updates Audio Element With New Props', () => {
      let wrapper = shallow(<TestAudioPlayerPresenter { ...testProps} />)

      let newProps = {
        ...testProps,
        audioPlayer: {
          muted: true,
          paused: false,
          volume: .5,
        }
      }

      wrapper.setProps(newProps)
      expect(audioMockRef.play).toHaveBeenCalled()
      expect(audioMockRef.muted).toEqual(newProps.audioPlayer.muted)
      expect(audioMockRef.paused).toEqual(newProps.audioPlayer.paused)
      expect(audioMockRef.volume).toEqual(newProps.audioPlayer.volume)

      let nextProps = {
        ...newProps,
        audioPlayer: {
          paused: true,
          muted: false,
          volume: .75
        }
      }

      wrapper.setProps(nextProps)

      expect(audioMockRef.pause).toHaveBeenCalled()
      expect(audioMockRef.muted).toEqual(nextProps.audioPlayer.muted)
      expect(audioMockRef.paused).toEqual(nextProps.audioPlayer.paused)
      expect(audioMockRef.volume).toEqual(nextProps.audioPlayer.volume)

      let newTrackProps = {
        ...nextProps,
        audioPlayer: {
          currentTrackId: itemFixtures[1].id,
          currentTime: itemFixtures[1].attributes.playtime,
          canPlay: true
        },
        item: itemFixtures[1]
      }

      wrapper.setProps(newTrackProps)

      expect(audioMockRef.play).toHaveBeenCalled()
      expect(audioMockRef.src).toEqual(itemFixtures[1].attributes.audio_url)

      let forceTimeUpdate = {
        ...newTrackProps,
        audioPlayer: {updateAudioElementTime: true}
      }

      wrapper.setProps(forceTimeUpdate)
      expect(audioMockRef.currentTime).toEqual(forceTimeUpdate.audioPlayer.currentTime)
    })
  })

  describe('Redux Connection', () => {

    describe('Sets State', () => {
      let newPlayer = new AudioPlayerState(itemFixtures[0].id),
          newState = { audioPlayer: newPlayer },
          result = mapStateToProps(newState)

      expect(result.audioPlayer).toBe(newPlayer)
    })

    describe('Dispatches Action', () => {
      let dispatchSpy

      beforeEach(() => {
        dispatchSpy = jest.fn()
      })
      it('`audioCanPlay`', () => {
        let { audioCanPlay } = mapDispatchToProps(dispatchSpy)

        audioCanPlay()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(audioCanPlayAction())
      })

      it('`audioMetaDataLoaded`', () => {
        let { audioMetaDataLoaded } = mapDispatchToProps(dispatchSpy)

        audioMetaDataLoaded()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(audioMetaDataLoadedAction())
      })


      it('`playClick`', () => {
        let item = itemFixtures[0],
            { play } = mapDispatchToProps(dispatchSpy, {item: item})

        play()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(playClick(item.id))
      })

      it('`pauseClick`', () => {
        let item = itemFixtures[0],
            { pause } = mapDispatchToProps(dispatchSpy, {item: item})

        pause()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(pauseClick(item.id))
      })

      it('`muteClick`', () => {
        let { mute } = mapDispatchToProps(dispatchSpy)

        mute()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(muteClick())
      })

      it('`unmuteClick`', () => {
        let { unmute } = mapDispatchToProps(dispatchSpy)

        unmute()

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(unmuteClick())
      })

      it('`updatePlayTimeTimeKeeper`', () => {
        let item = itemFixtures[0],
            { updatePlayTimeTimeKeeper } = mapDispatchToProps(dispatchSpy, {item: item}),
            newTime = 100

        updatePlayTimeTimeKeeper(newTime)

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(updatePlayTime(item.id, newTime))
      })

      it('`updatePlayTimeTimeScrubber`', () => {
        let { updatePlayTimeTimeScrubber } = mapDispatchToProps(dispatchSpy, {item: itemFixtures[0]}),
            newTime = 100

        updatePlayTimeTimeScrubber(newTime)

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(updatePlayTime(itemFixtures[0].id, newTime))
      })

      it('`timeScrubberChange`', () => {
        let { timeScrubberChange } = mapDispatchToProps(dispatchSpy, {item: itemFixtures[0]}),
            newTime = 100

        timeScrubberChange(newTime)

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(timeScrubberChangeAction(itemFixtures[0].id, newTime))
      })

      it('`updateVolume`', () => {
        let { updateVolume } = mapDispatchToProps(dispatchSpy),
            newVol = .5,
            event = new Event('testevent')

        updateVolume(newVol)

        expect(dispatchSpy).toHaveBeenCalled()
        expect(dispatchSpy.mock.calls[0][0]).toEqual(volumeChange(newVol))
      })
    })

  })
})
