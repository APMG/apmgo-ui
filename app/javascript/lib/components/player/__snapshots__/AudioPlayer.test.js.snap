// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer Component Test Presenter Renders 1`] = `
<div
  className="AudioPlayer__player___55AvF"
>
  <h2
    className="invisible"
  >
    Now Playing
  </h2>
  <audio />
  <div
    className="AudioPlayer__progress___-QYfn"
  >
    <h3
      className="invisible"
    >
      Time Control
    </h3>
    <div />
    <div
      className="AudioPlayer__timestamps___3iXPU"
    >
      <span
        className="AudioPlayer__currentTime___2CUk7"
      >
        0:00
      </span>
      <span
        className="AudioPlayer__duration___2Ss-f"
      >
        0:00
      </span>
    </div>
    <div>
      <input
        aria-label="Time Scrubber"
        className="AudioPlayer__scrubber___2VRHx"
        max={undefined}
        min={0}
        onInput={[Function]}
        onMouseUp={[Function]}
        type="range"
        value={0}
      />
    </div>
  </div>
  <h3
    className="AudioPlayer__title___1hJbf"
  >
    Test Audio
  </h3>
  <div
    className="AudioPlayer__controls___1S5g_"
  >
    <button
      className="AudioPlayer__skipBack___1HFfH AudioPlayer__skipBtn___gkURa"
    >
      <svg
        height="35"
        viewBox="0 0 30 35"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <g
          fillRule="nonzero"
          stroke="none"
          strokeWidth="1"
        >
          <path
            d="M29.98 20.03C29.98 28.28 23.27 34.99 15.02 34.99 6.77 34.99 0.05 28.28 0.05 20.03 0.05 19.54 0.45 19.14 0.94 19.14 1.43 19.14 1.83 19.54 1.83 20.03 1.83 27.3 7.74 33.22 15.02 33.22 22.29 33.22 28.2 27.3 28.2 20.03 28.2 12.76 22.29 6.84 15.02 6.84L13.48 6.84 15.68 9.3C16 9.67 15.97 10.23 15.61 10.56 15.24 10.88 14.68 10.85 14.35 10.49L11.21 6.96C10.69 6.39 10.69 5.52 11.21 4.94L14.35 1.42C14.53 1.22 14.77 1.12 15.02 1.12 15.23 1.12 15.44 1.2 15.61 1.35 15.97 1.67 16 2.24 15.68 2.6L13.48 5.07 15.02 5.07C23.27 5.07 29.98 11.78 29.98 20.03Z"
          />
        </g>
      </svg>
      <span
        className="invisible"
      >
        Skip back 
      </span>
      15
      <span
        className="invisible"
      >
        seconds
      </span>
    </button>
    <button
      className="AudioPlayer__playPauseBtn___1BAhn"
      disabled={true}
      id="play-pause-loading-button"
    >
      Loading ...
    </button>
    <button
      className="AudioPlayer__skipFwd___eUhyE AudioPlayer__skipBtn___gkURa"
    >
      <svg
        height="35"
        viewBox="0 0 30 35"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <g
          fillRule="nonzero"
          stroke="none"
          strokeWidth="1"
        >
          <path
            d="M0.02 20.03C0.02 28.28 6.73 34.99 14.98 34.99 23.23 34.99 29.95 28.28 29.95 20.03 29.95 19.54 29.55 19.14 29.06 19.14 28.57 19.14 28.17 19.54 28.17 20.03 28.17 27.3 22.26 33.22 14.98 33.22 7.71 33.22 1.8 27.3 1.8 20.03 1.8 12.76 7.71 6.84 14.98 6.84L16.52 6.84 14.32 9.3C14 9.67 14.03 10.23 14.39 10.56 14.76 10.88 15.32 10.85 15.65 10.49L18.79 6.96C19.31 6.39 19.31 5.52 18.79 4.94L15.65 1.42C15.47 1.22 15.23 1.12 14.98 1.12 14.77 1.12 14.56 1.2 14.39 1.35 14.03 1.67 14 2.24 14.32 2.6L16.52 5.07 14.98 5.07C6.73 5.07 0.02 11.78 0.02 20.03Z"
          />
        </g>
      </svg>
      <span
        className="invisible"
      >
        Skip forward 
      </span>
      15
      <span
        className="invisible"
      >
        seconds
      </span>
    </button>
  </div>
  <div
    className="AudioPlayer__volume___1Kdd1"
  >
    <button
      id="mute-button"
      onClick={undefined}
    >
      Mute
    </button>
    <h3>
      Volume Control
    </h3>
    <input
      aria-label="Volume"
      max={1}
      min={0}
      onChange={[Function]}
      onInput={[Function]}
      step={0.01}
      type="range"
      value={1}
    />
  </div>
</div>
`;
